File: common/src/test/java/DahRealPanda/plugins/colonyweb/auth/WebUserTest.java
package DahRealPanda.plugins.colonyweb.auth;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * The account record itself: how the two sources of colony access combine, and when a stored
 * session stops counting.
 */
class WebUserTest {

    private static StoredSession session(long expiresAt) {
        return new StoredSession("hash-" + expiresAt, 0, expiresAt);
    }

    @Nested
    @DisplayName("accessible colonies")
    class AccessibleColonies {

        @Test
        @DisplayName("membership and grants are combined")
        void unionOfBoth() {
            WebUser user = new WebUser("uuid", "Alice");
            user.colonies = new java.util.LinkedHashSet<>(List.of(1, 2));
            user.granted = new java.util.LinkedHashSet<>(List.of(3));

            assertEquals(Set.of(1, 2, 3), user.accessibleColonies());
        }

        @Test
        @DisplayName("a colony that is both a membership and a grant appears once")
        void overlapIsDeduplicated() {
            WebUser user = new WebUser("uuid", "Alice");
            user.colonies = new java.util.LinkedHashSet<>(List.of(1));
            user.granted = new java.util.LinkedHashSet<>(List.of(1));

            assertEquals(1, user.accessibleColonies().size());
        }

        @Test
        @DisplayName("a brand new account can see nothing")
        void emptyByDefault() {
            assertTrue(new WebUser("uuid", "Alice").accessibleColonies().isEmpty());
        }

        @Test
        @DisplayName("the returned set is a copy, so callers cannot grant themselves access")
        void returnsCopy() {
            WebUser user = new WebUser("uuid", "Alice");
            user.colonies = new java.util.LinkedHashSet<>(List.of(1));

            user.accessibleColonies().add(99);

            assertEquals(Set.of(1), user.accessibleColonies());
        }
    }

    @Nested
    @DisplayName("purging sessions")
    class Purging {

        @Test
        @DisplayName("an expired session is dropped and the change is reported")
        void dropsExpired() {
            WebUser user = new WebUser("uuid", "Alice");
            user.sessions.add(session(50));

            assertTrue(user.purgeExpiredSessions(100));
            assertTrue(user.sessions.isEmpty());
        }

        @Test
        @DisplayName("a live session is kept and no change is reported")
        void keepsLive() {
            WebUser user = new WebUser("uuid", "Alice");
            user.sessions.add(session(500));

            assertFalse(user.purgeExpiredSessions(100));
            assertEquals(1, user.sessions.size());
        }

        @Test
        @DisplayName("a session expiring exactly now is dropped")
        void boundaryIsExpired() {
            WebUser user = new WebUser("uuid", "Alice");
            user.sessions.add(session(100));

            assertTrue(user.purgeExpiredSessions(100));
        }

        @Test
        @DisplayName("only the expired sessions go")
        void mixedSessions() {
            WebUser user = new WebUser("uuid", "Alice");
            user.sessions.add(session(50));
            user.sessions.add(session(500));

            assertTrue(user.purgeExpiredSessions(100));
            assertEquals(1, user.sessions.size());
            assertEquals(500, user.sessions.get(0).expiresAt);
        }

        @Test
        @DisplayName("purging an account with no sessions reports no change")
        void noSessions() {
            assertFalse(new WebUser("uuid", "Alice").purgeExpiredSessions(100));
        }
    }
}

File: common/src/test/kotlin/DahRealPanda/plugins/colonyweb/auth/WebUserTest.kt
package DahRealPanda.plugins.colonyweb.auth

import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.collections.shouldContainExactlyInAnyOrder
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.shouldBe

/**
 * [WebUser] keeps the two sources of access apart on purpose: the colonies mirrored from
 * MineColonies, and the ones an operator granted by hand. The union of the two is what the auth
 * guard actually checks, so it is worth testing on its own.
 */
class WebUserTest : DescribeSpec({

    fun sessionExpiringAt(expiresAt: Long) = StoredSession("hash-$expiresAt", 0L, expiresAt)

    describe("accessibleColonies") {
        it("combines mirrored and granted colonies") {
            val user = WebUser(colonies = mutableSetOf(1, 2), granted = mutableSetOf(7))

            user.accessibleColonies() shouldContainExactlyInAnyOrder listOf(1, 2, 7)
        }

        it("counts a colony that appears in both lists once") {
            val user = WebUser(colonies = mutableSetOf(1, 2), granted = mutableSetOf(2))

            user.accessibleColonies() shouldContainExactlyInAnyOrder listOf(1, 2)
        }

        it("is empty for a player with no access at all") {
            WebUser().accessibleColonies() shouldHaveSize 0
        }

        it("works when only one of the two sources has anything in it") {
            WebUser(granted = mutableSetOf(7)).accessibleColonies() shouldContainExactlyInAnyOrder listOf(7)
            WebUser(colonies = mutableSetOf(1)).accessibleColonies() shouldContainExactlyInAnyOrder listOf(1)
        }

        // The guard reads this on every request, so it must not be a live view onto the sets it
        // was built from — a caller mutating the result must not silently widen the player's access.
        it("returns a snapshot rather than a live view of the underlying sets") {
            val user = WebUser(colonies = mutableSetOf(1))
            val accessible = user.accessibleColonies()

            user.colonies.add(2)

            accessible shouldContainExactlyInAnyOrder listOf(1)
        }
    }

    describe("purgeExpiredSessions") {
        it("drops a session that has expired and says so") {
            val user = WebUser(sessions = mutableListOf(sessionExpiringAt(50L)))

            user.purgeExpiredSessions(100L) shouldBe true
            user.sessions shouldHaveSize 0
        }

        it("keeps a session that is still live and reports no change") {
            val user = WebUser(sessions = mutableListOf(sessionExpiringAt(500L)))

            user.purgeExpiredSessions(100L) shouldBe false
            user.sessions shouldHaveSize 1
        }

        it("drops only the expired sessions of a player with several browsers") {
            val user = WebUser(sessions = mutableListOf(sessionExpiringAt(50L), sessionExpiringAt(500L)))

            user.purgeExpiredSessions(100L) shouldBe true
            user.sessions.map { it.expiresAt } shouldContainExactlyInAnyOrder listOf(500L)
        }

        it("treats a session expiring exactly now as expired") {
            val user = WebUser(sessions = mutableListOf(sessionExpiringAt(100L)))

            user.purgeExpiredSessions(100L) shouldBe true
        }

        it("reports no change for a player who has never signed in") {
            WebUser().purgeExpiredSessions(100L) shouldBe false
        }
    }
})
